File: cmd/drift/internal/templates/ios/NativeAudioPlayer.swift
/// NativeAudioPlayer.swift
/// Provides audio-only playback using AVPlayer via a standalone platform channel.
/// Supports multiple concurrent player instances, each identified by a playerId.

import AVFoundation

// MARK: - Audio Player Instance

/// Per-instance audio player state.
private class AudioPlayerInstance {
    let id: Int
    let player: AVQueuePlayer
    private var timeObserver: Any?
    private var timeControlObservation: NSKeyValueObservation?
    private var itemStatusObservation: NSKeyValueObservation?
    private var endOfItemObserver: NSObjectProtocol?
    private var playerLooper: AVPlayerLooper?
    private var playbackSpeed: Float = 1.0
    private var isLooping: Bool = false
    private var hasReachedEnd: Bool = false
    private var isStopped: Bool = false

    init(id: Int) {
        self.id = id

        DriftMediaSession.activate()

        self.player = AVQueuePlayer()

        // Observe time control status for playback state
        timeControlObservation = player.observe(\.timeControlStatus) { [weak self] player, _ in
            guard let self = self else { return }
            let state: Int
            switch player.timeControlStatus {
            case .paused:
                if player.currentItem == nil || self.isStopped {
                    state = 0 // Idle
                } else if self.hasReachedEnd {
                    state = 3 // Completed
                } else {
                    state = 4 // Paused
                }
                self.stopPositionUpdates()
            case .waitingToPlayAtSpecifiedRate:
                state = 1 // Buffering
                self.stopPositionUpdates()
            case .playing:
                state = 2 // Playing
                self.startPositionUpdates()
            @unknown default:
                state = 0 // Idle
                self.stopPositionUpdates()
            }
            self.sendStateEvent(state: state)
        }
    }

    private func startPositionUpdates() {
        guard timeObserver == nil else { return }
        let interval = CMTime(seconds: 0.25, preferredTimescale: CMTimeScale(NSEC_PER_SEC))
        timeObserver = player.addPeriodicTimeObserver(forInterval: interval, queue: .main) { [weak self] time in
            guard let self = self else { return }
            let positionMs = time.isNumeric ? Int64(CMTimeGetSeconds(time) * 1000) : 0
            var durationMs: Int64 = 0
            var bufferedMs: Int64 = 0

            if let item = self.player.currentItem {
                if item.duration.isNumeric {
                    durationMs = Int64(CMTimeGetSeconds(item.duration) * 1000)
                }
                if let timeRange = item.loadedTimeRanges.last?.timeRangeValue {
                    let bufferedEnd = CMTimeAdd(timeRange.start, timeRange.duration)
                    bufferedMs = Int64(CMTimeGetSeconds(bufferedEnd) * 1000)
                }
            }

            PlatformChannelManager.shared.sendEvent(
                channel: "drift/audio_player/events",
                data: [
                    "playerId": self.id,
                    "playbackState": 2, // Playing
                    "positionMs": positionMs,
                    "durationMs": max(durationMs, 0),
                    "bufferedMs": bufferedMs
                ]
            )
        }
    }

    private func stopPositionUpdates() {
        if let observer = timeObserver {
            player.removeTimeObserver(observer)
            timeObserver = nil
        }
    }

    func sendStateEvent(state: Int) {
        let currentTime = player.currentTime()
        let positionMs = currentTime.isNumeric ? Int64(CMTimeGetSeconds(currentTime) * 1000) : 0
        var durationMs: Int64 = 0
        var bufferedMs: Int64 = 0

        if let item = player.currentItem {
            if item.duration.isNumeric {
                durationMs = Int64(CMTimeGetSeconds(item.duration) * 1000)
            }
            if let timeRange = item.loadedTimeRanges.last?.timeRangeValue {
                let bufferedEnd = CMTimeAdd(timeRange.start, timeRange.duration)
                bufferedMs = Int64(CMTimeGetSeconds(bufferedEnd) * 1000)
            }
        }

        PlatformChannelManager.shared.sendEvent(
            channel: "drift/audio_player/events",
            data: [
                "playerId": id,
                "playbackState": state,
                "positionMs": positionMs,
                "durationMs": max(durationMs, 0),
                "bufferedMs": bufferedMs
            ]
        )
    }

    func load(url: URL) {
        isStopped = false
        hasReachedEnd = false

        // Disable existing looper before replacing the item
        playerLooper?.disableLooping()
        playerLooper = nil

        // Remove previous end-of-item observer
        if let observer = endOfItemObserver {
            NotificationCenter.default.removeObserver(observer)
            endOfItemObserver = nil
        }

        let item = AVPlayerItem(url: url)
        player.replaceCurrentItem(with: item)

        // Observe item status for errors
        itemStatusObservation?.invalidate()
        itemStatusObservation = item.observe(\.status) { [weak self] item, _ in
            guard let self = self else { return }
            if item.status == .failed {
                PlatformChannelManager.shared.sendEvent(
                    channel: "drift/audio_player/errors",
                    data: [
                        "playerId": self.id,
                        "code": mediaErrorCode(for: item.error),
                        "message": item.error?.localizedDescription ?? "Unknown playback error"
                    ]
                )
            }
        }

        // Register end-of-item observer for completion detection.
        // When AVPlayerLooper is active it prevents this notification from firing,
        // so the observer and looper are naturally mutually exclusive.
        endOfItemObserver = NotificationCenter.default.addObserver(
            forName: .AVPlayerItemDidPlayToEndTime,
            object: item,
            queue: .main
        ) { [weak self] _ in
            guard let self = self else { return }
            self.hasReachedEnd = true
            self.sendStateEvent(state: 3) // Completed
        }

        // Re-create looper if looping is active
        if isLooping {
            playerLooper = AVPlayerLooper(player: player, templateItem: item)
        }
    }

    func play() {
        isStopped = false
        hasReachedEnd = false
        player.play()
        if playbackSpeed != 1.0 {
            player.rate = playbackSpeed
        }
    }

    func pause() {
        player.pause()
    }

    func stop() {
        isStopped = true
        hasReachedEnd = false
        player.pause()
        player.seek(to: .zero) { [weak self] _ in
            guard let self = self else { return }
            self.sendStateEvent(state: 0) // Idle with position reset to zero
        }
    }

    func seekTo(positionMs: Int64) {
        hasReachedEnd = false
        let time = CMTime(seconds: Double(positionMs) / 1000.0, preferredTimescale: CMTimeScale(NSEC_PER_SEC))
        player.seek(to: time)
    }

    func setVolume(_ volume: Float) {
        player.volume = volume
    }

    func setLooping(_ looping: Bool) {
        isLooping = looping

        // Disable existing looper
        playerLooper?.disableLooping()
        playerLooper = nil

        if looping, let item = player.currentItem {
            playerLooper = AVPlayerLooper(player: player, templateItem: item)
        }
    }

    func setPlaybackSpeed(_ rate: Float) {
        playbackSpeed = rate
        if player.timeControlStatus == .playing {
            player.rate = rate
        }
    }

    func dispose() {
        stopPositionUpdates()
        timeControlObservation?.invalidate()
        timeControlObservation = nil
        itemStatusObservation?.invalidate()
        itemStatusObservation = nil
        if let observer = endOfItemObserver {
            NotificationCenter.default.removeObserver(observer)
            endOfItemObserver = nil
        }
        playerLooper?.disableLooping()
        playerLooper = nil
        player.pause()
        player.replaceCurrentItem(with: nil)
        playbackSpeed = 1.0
        isLooping = false
        hasReachedEnd = false
        isStopped = false
    }
}

// MARK: - Audio Player Handler

/// Handles audio player platform channel methods from Go.
/// Manages multiple player instances keyed by playerId.
enum AudioPlayerHandler {
    private static var players: [Int: AudioPlayerInstance] = [:]

    static func handle(method: String, args: Any?) -> (Any?, Error?) {
        let argsMap = args as? [String: Any]
        let playerId = (argsMap?["playerId"] as? NSNumber)?.intValue ?? 0

        switch method {
        case "load":
            return load(playerId: playerId, args: argsMap)
        case "play":
            return play(playerId: playerId)
        case "pause":
            return pause(playerId: playerId)
        case "stop":
            return stop(playerId: playerId)
        case "seekTo":
            return seekTo(playerId: playerId, args: argsMap)
        case "setVolume":
            return setVolume(playerId: playerId, args: argsMap)
        case "setLooping":
            return setLooping(playerId: playerId, args: argsMap)
        case "setPlaybackSpeed":
            return setPlaybackSpeed(playerId: playerId, args: argsMap)
        case "dispose":
            return dispose(playerId: playerId)
        default:
            return (nil, NSError(domain: "AudioPlayer", code: 404, userInfo: [NSLocalizedDescriptionKey: "Unknown method: \(method)"]))
        }
    }

    private static func ensurePlayer(playerId: Int) -> AudioPlayerInstance {
        if let existing = players[playerId] {
            return existing
        }
        let instance = AudioPlayerInstance(id: playerId)
        players[playerId] = instance
        return instance
    }

    private static func load(playerId: Int, args: [String: Any]?) -> (Any?, Error?) {
        guard let url = args?["url"] as? String,
              let mediaURL = URL(string: url) else {
            return (nil, NSError(domain: "AudioPlayer", code: 400, userInfo: [NSLocalizedDescriptionKey: "Missing url"]))
        }

        ensurePlayer(playerId: playerId).load(url: mediaURL)
        return (nil, nil)
    }

    private static func play(playerId: Int) -> (Any?, Error?) {
        ensurePlayer(playerId: playerId).play()
        return (nil, nil)
    }

    private static func pause(playerId: Int) -> (Any?, Error?) {
        players[playerId]?.pause()
        return (nil, nil)
    }

    private static func stop(playerId: Int) -> (Any?, Error?) {
        players[playerId]?.stop()
        return (nil, nil)
    }

    private static func seekTo(playerId: Int, args: [String: Any]?) -> (Any?, Error?) {
        let positionMs = args?["positionMs"] as? Int64 ?? 0
        players[playerId]?.seekTo(positionMs: positionMs)
        return (nil, nil)
    }

    private static func setVolume(playerId: Int, args: [String: Any]?) -> (Any?, Error?) {
        let volume = (args?["volume"] as? NSNumber)?.floatValue ?? 1.0
        players[playerId]?.setVolume(volume)
        return (nil, nil)
    }

    private static func setLooping(playerId: Int, args: [String: Any]?) -> (Any?, Error?) {
        let looping = args?["looping"] as? Bool ?? false
        players[playerId]?.setLooping(looping)
        return (nil, nil)
    }

    private static func setPlaybackSpeed(playerId: Int, args: [String: Any]?) -> (Any?, Error?) {
        let rate = (args?["rate"] as? NSNumber)?.floatValue ?? 1.0
        players[playerId]?.setPlaybackSpeed(rate)
        return (nil, nil)
    }

    private static func dispose(playerId: Int) -> (Any?, Error?) {
        players.removeValue(forKey: playerId)?.dispose()
        DriftMediaSession.deactivate()
        return (nil, nil)
    }
}

File: cmd/drift/internal/templates/android/java/NativeAudioPlayer.kt
/**
 * NativeAudioPlayer.kt
 * Provides audio-only playback using ExoPlayer via a standalone platform channel.
 * Supports multiple concurrent player instances, each identified by a playerId.
 */
package {{.PackageName}}

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.media3.common.AudioAttributes
import androidx.media3.common.C
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.exoplayer.ExoPlayer

/**
 * Per-instance audio player state.
 */
private class AudioPlayerInstance(
    val id: Long,
    context: Context,
    private val handler: Handler
) {
    val player: ExoPlayer = ExoPlayer.Builder(context).build().also {
        it.setAudioAttributes(
            AudioAttributes.Builder()
                .setUsage(C.USAGE_MEDIA)
                .setContentType(C.AUDIO_CONTENT_TYPE_MUSIC)
                .build(),
            /* handleAudioFocus= */ true
        )
    }
    private var positionRunnable: Runnable? = null

    init {
        player.addListener(object : Player.Listener {
            override fun onPlaybackStateChanged(playbackState: Int) {
                val state = when (playbackState) {
                    Player.STATE_IDLE -> 0
                    Player.STATE_BUFFERING -> 1
                    Player.STATE_READY -> if (player.isPlaying) 2 else 4
                    Player.STATE_ENDED -> {
                        stopPositionUpdates()
                        3
                    }
                    else -> 0
                }
                sendStateEvent(state)
            }

            override fun onIsPlayingChanged(isPlaying: Boolean) {
                if (player.playbackState == Player.STATE_READY) {
                    val state = if (isPlaying) 2 else 4 // Playing or Paused
                    sendStateEvent(state)
                    if (isPlaying) startPositionUpdates() else stopPositionUpdates()
                }
            }

            override fun onPlayerError(error: PlaybackException) {
                PlatformChannelManager.sendEvent(
                    "drift/audio_player/errors",
                    mapOf(
                        "playerId" to id,
                        "code" to mediaErrorCodeString(error.errorCode),
                        "message" to (error.message ?: "Unknown playback error")
                    )
                )
            }
        })
    }

    fun sendStateEvent(state: Int) {
        PlatformChannelManager.sendEvent(
            "drift/audio_player/events",
            mapOf(
                "playerId" to id,
                "playbackState" to state,
                "positionMs" to player.currentPosition,
                "durationMs" to player.duration.coerceAtLeast(0),
                "bufferedMs" to player.bufferedPosition
            )
        )
    }

    fun startPositionUpdates() {
        stopPositionUpdates()
        positionRunnable = object : Runnable {
            override fun run() {
                if (player.playbackState != Player.STATE_IDLE) {
                    PlatformChannelManager.sendEvent(
                        "drift/audio_player/events",
                        mapOf(
                            "playerId" to id,
                            "playbackState" to when (player.playbackState) {
                                Player.STATE_IDLE -> 0
                                Player.STATE_BUFFERING -> 1
                                Player.STATE_READY -> if (player.isPlaying) 2 else 4
                                Player.STATE_ENDED -> 3
                                else -> 0
                            },
                            "positionMs" to player.currentPosition,
                            "durationMs" to player.duration.coerceAtLeast(0),
                            "bufferedMs" to player.bufferedPosition
                        )
                    )
                }
                handler.postDelayed(this, 250)
            }
        }
        handler.post(positionRunnable!!)
    }

    fun stopPositionUpdates() {
        positionRunnable?.let { handler.removeCallbacks(it) }
        positionRunnable = null
    }

    fun dispose() {
        stopPositionUpdates()
        player.release()
    }
}

/**
 * Handles audio player platform channel methods from Go.
 * Manages multiple player instances keyed by playerId.
 */
object AudioPlayerHandler {
    private var context: Context? = null
    private val handler = Handler(Looper.getMainLooper())
    private val players = mutableMapOf<Long, AudioPlayerInstance>()

    fun handle(context: Context, method: String, args: Any?): Pair<Any?, Exception?> {
        if (this.context == null) {
            this.context = context.applicationContext
        }
        val argsMap = args as? Map<*, *>
        val playerId = (argsMap?.get("playerId") as? Number)?.toLong() ?: 0L

        return when (method) {
            "load" -> load(playerId, argsMap)
            "play" -> play(playerId)
            "pause" -> pause(playerId)
            "stop" -> stop(playerId)
            "seekTo" -> seekTo(playerId, argsMap)
            "setVolume" -> setVolume(playerId, argsMap)
            "setLooping" -> setLooping(playerId, argsMap)
            "setPlaybackSpeed" -> setPlaybackSpeed(playerId, argsMap)
            "dispose" -> dispose(playerId)
            else -> Pair(null, IllegalArgumentException("Unknown method: $method"))
        }
    }

    private fun ensurePlayer(playerId: Long): AudioPlayerInstance {
        return players.getOrPut(playerId) {
            val ctx = context ?: throw IllegalStateException("Context not initialized")
            AudioPlayerInstance(playerId, ctx, handler)
        }
    }

    private fun load(playerId: Long, args: Map<*, *>?): Pair<Any?, Exception?> {
        val url = args?.get("url") as? String
            ?: return Pair(null, IllegalArgumentException("Missing url"))

        handler.post {
            val instance = ensurePlayer(playerId)
            val mediaItem = MediaItem.fromUri(url)
            instance.player.setMediaItem(mediaItem)
            instance.player.prepare()
        }
        return Pair(null, null)
    }

    private fun play(playerId: Long): Pair<Any?, Exception?> {
        handler.post {
            val player = ensurePlayer(playerId).player
            if (player.playbackState == Player.STATE_IDLE) {
                player.prepare()
            }
            player.play()
        }
        return Pair(null, null)
    }

    private fun pause(playerId: Long): Pair<Any?, Exception?> {
        handler.post {
            players[playerId]?.player?.pause()
        }
        return Pair(null, null)
    }

    private fun stop(playerId: Long): Pair<Any?, Exception?> {
        handler.post {
            val instance = players[playerId] ?: return@post
            instance.player.stop()
            instance.player.seekTo(0)
            instance.sendStateEvent(0) // Idle with position reset to zero
        }
        return Pair(null, null)
    }

    private fun seekTo(playerId: Long, args: Map<*, *>?): Pair<Any?, Exception?> {
        val positionMs = (args?.get("positionMs") as? Number)?.toLong() ?: 0L
        handler.post {
            players[playerId]?.player?.seekTo(positionMs)
        }
        return Pair(null, null)
    }

    private fun setVolume(playerId: Long, args: Map<*, *>?): Pair<Any?, Exception?> {
        val volume = (args?.get("volume") as? Number)?.toFloat() ?: 1.0f
        handler.post {
            players[playerId]?.player?.volume = volume
        }
        return Pair(null, null)
    }

    private fun setLooping(playerId: Long, args: Map<*, *>?): Pair<Any?, Exception?> {
        val looping = args?.get("looping") as? Boolean ?: false
        handler.post {
            players[playerId]?.player?.repeatMode = if (looping) Player.REPEAT_MODE_ALL else Player.REPEAT_MODE_OFF
        }
        return Pair(null, null)
    }

    private fun setPlaybackSpeed(playerId: Long, args: Map<*, *>?): Pair<Any?, Exception?> {
        val rate = (args?.get("rate") as? Number)?.toFloat() ?: 1.0f
        handler.post {
            players[playerId]?.player?.setPlaybackSpeed(rate)
        }
        return Pair(null, null)
    }

    private fun dispose(playerId: Long): Pair<Any?, Exception?> {
        handler.post {
            players.remove(playerId)?.dispose()
        }
        return Pair(null, null)
    }
}
